plugins/destination/managed_writer.go

```diff
@@ -0,0 +1,162 @@
+package destination
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/cloudquery/plugin-sdk/schema"
+	"github.com/cloudquery/plugin-sdk/specs"
+)
+
+type worker struct {
+	count int
+	wg    *sync.WaitGroup
+	ch    chan schema.CQTypes
+	flush chan chan bool
+}
+
+func (p *Plugin) worker(ctx context.Context, metrics *Metrics, table *schema.Table, ch <-chan schema.CQTypes, flush <-chan chan bool) {
+	resources := make([][]any, 0)
+	for {
+		select {
+		case r, ok := <-ch:
+			//nolint:revive
+			if ok {
+				resources = append(resources, schema.TransformWithTransformer(p.client, r))
+				if len(resources) == p.spec.BatchSize {
+					start := time.Now()
+					if err := p.client.WriteTableBatch(ctx, table, resources); err != nil {
+						p.logger.Err(err).Str("table", table.Name).Int("len", p.spec.BatchSize).Dur("duration", time.Since(start)).Msg("failed to write batch")
+						// we don't return as we need to continue until channel is closed otherwise there will be a deadlock
+						atomic.AddUint64(&metrics.Errors, uint64(p.spec.BatchSize))
+					} else {
+						p.logger.Info().Str("table", table.Name).Int("len", p.spec.BatchSize).Dur("duration", time.Since(start)).Msg("batch written successfully")
+						atomic.AddUint64(&metrics.Writes, uint64(p.spec.BatchSize))
+					}
+					resources = make([][]interface{}, 0)
+				}
+			} else {
+				if len(resources) > 0 {
+					start := time.Now()
+					if err := p.client.WriteTableBatch(ctx, table, resources); err != nil {
+						p.logger.Err(err).Str("table", table.Name).Int("len", len(resources)).Dur("duration", time.Since(start)).Msg("failed to write last batch")
+						atomic.AddUint64(&metrics.Errors, uint64(len(resources)))
+					} else {
+						p.logger.Info().Str("table", table.Name).Int("len", len(resources)).Dur("duration", time.Since(start)).Msg("last batch written successfully")
+						atomic.AddUint64(&metrics.Writes, uint64(len(resources)))
+					}
+				}
+				return
+			}
+		case <-time.After(p.batchTimeout):
+			if len(resources) > 0 {
+				start := time.Now()
+				if err := p.client.WriteTableBatch(ctx, table, resources); err != nil {
+					p.logger.Err(err).Str("table", table.Name).Int("len", len(resources)).Dur("time", time.Since(start)).Msg("failed to write batch on timeout")
+					// we don't return as we need to continue until channel is closed otherwise there will be a deadlock
+					atomic.AddUint64(&metrics.Errors, uint64(len(resources)))
+				} else {
+					p.logger.Info().Str("table", table.Name).Int("len", len(resources)).Dur("time", time.Since(start)).Msg("batch written successfully on timeout")
+					atomic.AddUint64(&metrics.Writes, uint64(len(resources)))
+				}
+				resources = make([][]interface{}, 0)
+			}
+		case done := <-flush:
```

Comment:
because write can be called multiple times so we need to wait for a flush from a specific worker and we can't close the channel as other worker might be still using this worker, so we just wanted to do a flush.